src/java_bytecode/java_object_factory.cpp

```diff
@@ -189,20 +194,23 @@ void java_object_factoryt::gen_nondet_init(
     {
       const struct_typet &struct_type=to_struct_type(subtype);
       const irep_idt struct_tag=struct_type.get_tag();
-      // set to null if found in recursion set and not a sub-type
-      if(recursion_set.find(struct_tag)!=recursion_set.end() &&
-         struct_tag==class_identifier)
+      // set to null if found in recursion set and recursion depth allows it
+      if(recursion_set.find(struct_tag)!=recursion_set.end())
       {
-        // make null
-        null_pointer_exprt null_pointer_expr(pointer_type);
-        code_assignt code(expr, null_pointer_expr);
-        code.add_source_location()=loc;
-        init_code.copy_to_operands(code);
-
-        return;
+        if(recursion_depth<max_recursion_depth)
+          recursion_depth++;
```

Comment:
There is a problem with `null` initialization of objects. In the original implementation, a recursion set was used that would in an object hierarchy prevent instantiation of object types already seen. There were cases where this was problematic, if an object contained a recursive instantation, effectively this lead to a `null` value and due to various options (`--assertions-as-assumptions` etc.) created an `ASSUME non null` for that exact member, leading to inconsistent model.

The idea of this parameter was to create a solution between using exclusively the recursion set and not using it at all, i.e., having a bounded initialization of hierarchical objects.